pkg/analyze/data_source_test.go

```diff
@@ -0,0 +1,265 @@
+package analyze
+
+import (
+	"context"
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	"github.com/scylladb/scylla-operator/pkg/client/scylla/clientset/versioned/fake"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubefake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+	"reflect"
+	"testing"
+)
+
+type dummyLister struct {
+	indexer cache.Indexer
+}
+
+type dummyObject struct {
+	metav1.ObjectMeta
+	id int64
+}
+
+func (d dummyObject) GetObjectKind() schema.ObjectKind {
+	return schema.EmptyObjectKind
+}
+
+func (d dummyObject) DeepCopyObject() runtime.Object {
+	return d
+}
+
+func (d *dummyLister) List() ([]*dummyObject, error) {
+	do := make([]*dummyObject, 0)
+	for _, obj := range d.indexer.List() {
+		do = append(do, obj.(*dummyObject))
+	}
+	return do, nil
+}
+
+type buildListerTest[T runtime.Object, U any] struct {
+	ListerFactory func(i cache.Indexer) U
+	Objects       []*T
+}
+
+func assertExpectedContainsPermutatedActual[T, U any](t *testing.T, expectedResources []T, actualResources []U) {
+	if len(expectedResources) != len(actualResources) {
+		t.Errorf("expected %d resources, got %d", len(expectedResources), len(actualResources))
+		t.Logf("expected %v got %v", expectedResources, actualResources)
+	}
+
+	for _, expected := range expectedResources {
+		found := false
+		for _, actual := range actualResources {
+			if reflect.DeepEqual(actual, expected) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			t.Fatalf("object is missing: %v", expected)
+		}
+	}
+}
+
+func TestBuildLister(t *testing.T) {
+	buildListerDummyTests := map[string]buildListerTest[dummyObject, dummyLister]{
```

Comment:
nit: inline the unnamed struct, it's easier to read

Reply:
done